service/storage_proxy.cc

```diff
@@ -3990,16 +4012,17 @@ storage_proxy::query_singular(lw_shared_ptr<query::read_command> cmd,
                 std::pair<::shared_ptr<abstract_read_executor>, dht::token_range>& executor_and_token_range) {
         auto& rex = std::get<0>(executor_and_token_range);
         auto& token_range = std::get<1>(executor_and_token_range);
-        utils::latency_counter lc;
-        lc.start();
-        return rex->execute(timeout).then_wrapped([p = std::move(p), lc, rex, used_replicas, token_range = std::move(token_range), tmptr] (
+        return rex->execute(timeout).then_wrapped([p = std::move(p), rex, used_replicas, token_range = std::move(token_range), tmptr] (
                     future<foreign_ptr<lw_shared_ptr<query::result>>> f) mutable {
             if (!f.failed()) {
                 used_replicas->emplace(std::move(token_range), endpoints_to_replica_ids(*tmptr, rex->used_targets()));
             }
-            if (lc.is_start()) {
-                rex->get_cf()->add_coordinator_read_latency(lc.stop().latency());
+
+            auto latency = rex->max_request_latency();
+            if (latency) {
+                rex->get_cf()->add_coordinator_read_latency(*latency);
```

Comment:
Isn't this patch practically equivalent to:

```
if (!f.failed() && !rex->speculated()) {
    rex->get_cf()->add_coordinator_read_latency(lc.stop().latency());
}
```

Where `speculated()` returns if there was a speculation?

Reply:
Theoretically no. Practically I don't know.

The proposed alternative would cause us to potentially stop measuring latencies of majority of requests during periods of frequent speculative retries, potentially causing the metric to deteriorate during this time.

For example, suppose that our replicas are overloaded and they sometimes don't answer to coordinator requests, causing most requests to require a speculative retry to finish. The round-trip latencies of replica requests have risen due to the overload. So the 99th percentile latency has also risen - not because of speculative retry but because of overload. However, because most requests require a speculative retry, this solution would cause us not to measure them. So for a time longer than necessary we would keep using data points from before this period, i.e. we would execute speculative retries faster than we should (because our 99th percentile metric did not change as quickly as it should), contributing to the overload.

I think the safer solution is to keep measuring latencies for speculative retries as well.

Reply:
Not all requests use slow replica, so you still have up-to-date latency expectation. On the bright side you take clock measurement only once per request instead of potentially RF.

Reply:
I was thinking about a situation where the entire cluster is overloaded -> all replicas are slow -> almost all requests require a speculative retry -> the number of fresh data points is small -> the data points measuring the current latency remain insignificant compared to the previous measurements for longer than necessary.

The number of clock measurements is right now RF, but I'm gonna apply the review suggestions and they'll drop to 1 or 2 (2 only if there is a speculative retry).

Reply:
Actually it will be 1 measurement for all data read requests and 1 for all digest requests and then potentially one for another batch of data read requests if we decide to read repair and potentially 1 for speculative retry. If I'm not mistaken.
So from 1 to 4, with 2 being the common case (AFAIK read requests usually go with 1 data read + a bunch of digest reads).

Reply:
If all cluster becomes so overloaded that latency jumps to 99 percentile for all requests that may happen. But consider what will happen with your patch if most of the requests will fail because of network latencies become huge? The patch still updates read latency statistics with the latency of a local read, so the 99 percentile will go down instead of up.

Reply:
> Actually it will be 1 measurement for all data read requests and 1 for all digest requests and then potentially one for another batch of data read requests if we decide to read repair and potentially 1 for speculative retry. If I'm not mistaken.

What this means is that if a lot of read repairs are needed we will always speculate because the measurement only measures network latency now and does not take repair into account.

Reply:
> But consider what will happen with your patch if most of the requests will fail because of network latencies become huge? The patch still updates read latency statistics with the latency of a local read, so the 99 percentile will go down instead of up.

But requests which managed to read only from the local replica won't be measured because they'll fail, no? (this patch only measures successful requests)

Reply:
> > But consider what will happen with your patch if most of the requests will fail because of network latencies become huge? The patch still updates read latency statistics with the latency of a local read, so the 99 percentile will go down instead of up.
> 
> But requests which managed to read only from the local replica won't be measured because they'll fail, no? (this patch only measures successful requests)

I do not see that.

Reply:
Ah, indeed it measures all requests :/ and I wrote in the commit message that it measures only successful ones.
Must have lost that property at some point when refactoring. Will fix

Reply:
> > > But consider what will happen with your patch if most of the requests will fail because of network latencies become huge? The patch still updates read latency statistics with the latency of a local read, so the 99 percentile will go down instead of up.
> > 
> > 
> > But requests which managed to read only from the local replica won't be measured because they'll fail, no? (this patch only measures successful requests)
> 
> I do not see that.

Also it is not clear why they shouldn't be tacked into account. In will cause a lot of speculations in case two replicas are overloaded. Though it may not be a new issue since IIRC there is a cutoff for speculation around one second anyway.

Reply:
> Also it is not clear why they shouldn't be tacked into account.

It's not totally clear to me either. But I think that's just how the 99perenctile setting should work conceptually.
Why do we perform the retry after the request's duration hits the 99th percentile? Because we expected it to _succeed_ within this 99th percentile. It didn't, but we hope that only one replica is missing so we speculate.
In other words: it usually succeeds within `T`, so if it doesn't succeed within `T`, speculate.

Now the question is whether the 99percentile setting makes sense at all; perhaps it's just better for the user to use a constant threshold anyway?

---

I'm a bit worried about the read repair issue, but I don't know if it's a problem. Ok, if there's many read repairs we would speculate a bit more but I wonder whether such periods of many read repairs are common at all (these read repairs would need to take a majority of requests), and even if we do speculate more for some time - whether it would be a problem in this case (the speculative retries would still be rare assuming that the cluster is not overloaded at the same time).

I guess it's either that or the issue I mentioned. Not sure which is worse.

Well... I guess I could write a solution that solves both issues: keep the maximum of read-repair-data-requests measurements separate from the maximum of original data/digest requests (including speculation, if it happened), then sum the two maxes to get the final measurement. But that would introduce a bunch of complexity while it's not really clear if it's needed at all (I'd have to come up with an experiment which creates a data set where many read repairs are required and then actually confirm that it helps, and so on...)

Reply:
Actually now I think that what this PR does now makes more sense than if we extended the measurement by read repair duration.

A read that executes read repair contains two batches of requests: one for the initial data+digest reads, another for retrieving additional data for read repair.

Speculative retry is there to increase chances for _a single batch to succeed_ - the initial batch.

So it doesn't make sense to take the duration Y - X, where
1. measure X
2. one batch executes
3. the second batch executes (for read repair)
4. measure X

I think it makes more sense to take Y-X for
1. X
2. one batch executes
3. Y,

then Y-X for
1. X
2. second batch executes
3. Y,

(both adjusted to remove the feedback loop) and then take the maximum.
Which is equivalent to what this patch does.

Strictly speaking that second duration (for the second batch) is not needed and maybe it would be even better not to measure it at all, but I'm guessing it doesn't hurt as the requests are very similar to the requests made in the first batch (so their latencies should also be similar) and it makes the code a bit more consistent.